midend/interpreter.cpp

```diff
@@ -1124,11 +1153,17 @@ void ExpressionEvaluator::postorder(const IR::MethodCallExpression* expression)
         } else {
             BUG_CHECK(name == IR::Type_Header::isValid,
                       "%1%: unexpected method", bim->name);
-            BUG_CHECK(base->is<SymbolicHeader>(), "%1%: expected a header", base);
-            auto hv = base->to<SymbolicHeader>();
-            auto v = hv->valid;
-            set(expression, v);
-            return;
+            if (auto hv = structVar->to<SymbolicHeader>()) {
+                auto v = hv->valid;
+                set(expression, v);
+                return;
+            } else if (auto huv = structVar->to<SymbolicHeaderUnion>()) {
+                auto v = huv->valid;
```

Comment:
Considering that validity of a header union is dependent on its members, you may just want to check if one member is valid. If multiple members are valid that should be an assertion error.

Reply:
I do not quite understand how this relates to this part of the code, but here I just set the received value

Reply:
This is a method call on isValid right? A union is actually valid according to its members (https://p4.org/p4-spec/docs/P4-16-v1.2.2.html#sec-expr-hu). It technically does not have a validity bit. This relates to the other part of the comments, which ask whether we even need a valid field for a header union.